### ISSUE TITLE
Add waitRules for Custom Resources that don't have the "conditions" field in their status. 

### ISSUE THREAD

**Issue by mohansitaram** [k14s/kapp#442]
**Describe the problem/challenge you have**
kapp's waitRules currently only supports conditionMatchers for resources that have the "Conditions" field in their status. We are using kapp to deploy a Custom Resource that has no "Conditions" field in its status so we are not able to detect if the CR actually got deployed successfully. The CR does have a Status field though which could be used for adding waitRules

**Describe the solution you'd like**
A waitRules configuration that allows users to configure arbitrary rules. For example if the CR has this in its status field:
`status:
  currentState: Running
`
the conditionMatchers could be 
`- type: currentState
     statusMessage: Running
`
The above is just a suggestion but because this could be arbitrary, the conditionMatcher may need to support more complex data structures. 

**Anything else you would like to add:**

---
Vote on this request

This is an invitation to the community to vote on issues, to help us prioritize our backlog. Use the "smiley face" up to the right of this comment to vote.

👍 "I would like to see this addressed as soon as possible"
👎 "There are other more important things to focus on right now"

We are also happy to receive and review Pull Requests if you want to help working on this issue.

**Comment by renuy** [comment]
@mohansitaram,
Thanks for creating this issue. adding it to our backlog. We will try and get it to it soon.

ps:related to #11

**Comment by rohitagg2020** [comment]
Design Proposal: https://hackmd.io/hfgsE5v2QoyDjyu3VD0H4g

**Comment by sethiyash** [comment]
Acceptance Criteria - 
- User should be able to configure arbitrary waitRules with the help of a function with predeclared name (eg. is_done) in starlark. Below is the example where we are adding waitRules for the Custom Resource which have `status.currentState` -
```
apiVersion: kapp.k14s.io/v1alpha1
kind: Config

waitRules:
  - ytt:
      funcContractV1:
        resource.star: |
          def is_done(resource):
            state = resource.status.currentState
            if state == "Failed":
              return {"done": True, "successful": False, "message": ""}
            elif state == "Running":
              return {"done": True, "successful": True, "message": ""}
            else:
              return {"done":False, "successful": False, "message": "Not in Failed or Running state"}
            end
          end
  resourceMatchers:
    - apiVersionKindMatcher: {apiVersion: <resource-api-version>, kind: <resource-kind>}
```
- Kapp should internally use ytt package to call that function and get the result.

**Comment by jdef** [comment]
prior art: https://github.com/kubernetes/kubernetes/pull/105776

**Comment by renuy** [comment]
available in v0.48.0
